docs/src/index.md

```diff
@@ -118,21 +118,67 @@ applies the chain rule to go from `∂c/∂b` to `∂c/∂a`.
 The pushforward of `g`,  which also incorporates the knowledge of `∂b/∂a`,
 applies the chain rule to go from `∂a/∂x` to `∂b/∂x`.
 
-### Heavier Math
-If I have some functions: ``g(a)``, ``h(b)`` and ``f(x)=g(h(x))``, and I know
-the pullback of ``g``, at ``h(x)`` written: ``\mathrm{pullback}_{g(a)|a=h(x)}``,
-and I know the derivative of ``h`` with respect to its input ``b`` at ``g(x)``,
-written: ``\left.\dfrac{∂h}{∂b}\right|_{b=g(x)}`` Then I can use the pullback to
-find: ``\dfrac{∂f}{∂x}``:
+##### Geometric interpretation of reverse and forwards mode AD
 
-``\dfrac{∂f}{∂x}=\mathrm{\mathrm{pullback}_{g(a)|a=h(x)}}\left(\left.\dfrac{∂h}{∂b}\right|_{b=g(x)}\right).``
+Let us think of our types geometrically. In other words, elements of a type form a _manifold_.
+This document will explain this point of view in some detail.
 
-If I know the derivative of ``g`` with respect to its input a at ``x``, written:
-``\left.\dfrac{∂g}{∂a}\right|_{a=x}``, and I know the pushforward of ``h`` at
-``g(x)`` written: ``\mathrm{pushforward}_{h(b)|b=g(x)}``. Then I can use the
-pushforward to find ``\dfrac{∂f}{∂x}``:
+###### Some terminology/conventions.
 
-``\dfrac{∂f}{∂x}=\mathrm{pushforward}_{h(b)|b=g(x)}\left(\left.\dfrac{∂g}{∂a}\right|_{a=x}\right)``
+Let ``p`` be an element of type M, which is defined by some assignment of numbers ``x_1,...,x_m``,
+say ``(x_1,...,x_m) = (a_1,...,1_m)`` 
+
+A _function_ ``f:M -> K`` on ``M`` is (for simplicity) a polynomial ``K[x_1, ... x_m]``
+
+The tangent space ``T_pM`` of ``T`` at point ``p`` is the ``K``-vector space spanned by derivations ``d/dx``. 
+The tangent space acts linearly on the space of functions. They act as usual on functions. Our starting point is 
+that we know how to write down ``d/dx(f) = df/dx``.
+
+The collection of tangent spaces ``{T_pM}`` for ``p\in M`` is called the _tangent bundle_ of ``M``.
+
+Let ``df`` denote the first order information of ``f`` at each point. This is called the differential of ``f``. 
+If the derivatives of ``f`` and ``g`` agree at ``p``, we say that ``df`` and ``dg`` represent the same cotangent at ``p``.
+The covectors ``dx_1, ..., dx_m`` form the basis of the cotangent space T^*_pM at ``p``. Notice that this vector space is 
+dual to ``T_p``
+
+The collection of cotangent spaces ``{T^*_pM}`` for ``p\in M`` is called the _cotangent bundle_ of ``M``.
+
+###### Push-forwards and pullbacks
+
+Let ``N`` be another type, defined by numbers ``y_1,...,y_n``, and let ``g:M -> N`` be a _map_, that is, 
+an ``n``-dimensional vector ``(g_1, ..., g_m)`` of functions on ``M``.
+
+We define the _push-forward_ ``g_*:TM -> TN`` between tangent bundles by ``g_*(X)(h) = X(g\circ h)`` for any tangent vector ``X`` and function ``f``.
+We have ``g_*(d/dx_i)(y_j) = dg_j/dx_i, so the push-forward corresponds to the Jacobian, given a chosen basis.
+
+Similarly, the pullback of the differential ``df`` is defined by
+``g^*(df) = d(g\circ f)``. So for a coordinate differential ``dy_j``, we have
```

Comment:
Shouldn't the `g `f` order be opposite here. I.e `g` first maps M->N then, and  f:N->R , ``g^*(df) = d(f\circ g)``